airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py

```diff
@@ -126,5 +126,15 @@ def format_datetime(dt: Union[str, datetime.datetime], format: str) -> str:
     return dt_datetime.strftime(format)
 
 
-_macros_list = [now_utc, today_utc, timestamp, max, day_delta, duration, format_datetime]
+def format_datetime_string(dt: str, format: str) -> datetime.datetime:
```

Comment:
In which case would we want the interpolation to generate a datetime instead of a string?

Reply:
This is to handle this transformation from low code:
https://github.com/airbytehq/airbyte/blob/9daef0f143d354bd71c1c56777edc094fdefe1e9/airbyte-integrations/connectors/source-twilio/source_twilio/streams.py#L101

at manifest, I will do this:

```
transformations:
        - type: AddFields
          fields:
            - type: AddedFieldDefinition
              path:
                - date_created
              value: '{% set datetime_rfc2822 = format_datetime_string(record["date_created"], "%a, %d %b %Y %H:%M:%S %z") %}{% set formatted_datetime = format_datetime(datetime_rfc2822, "%Y-%m-%dT%H:%M:%SZ") %}{{ formatted_datetime }}'
```

On this way, we can convert any formatted string to datetime and then use format_datetime to make it an actual string.

Reply:
This seems like a complex path. Is there a way that format_datetime_string can handle all the processes instead of chaining results? Also, it feels like something should be very well documented.

Reply:
I would agree with Aldo's comment. It seems like what we want to do is to expand the capabilities of `format_datetime` to have any format as an input when it is a string and not just `isoparse` it. Could we modify this one to have something like:

```
def format_datetime(dt: Union[str, datetime.datetime], output_format: str, input_format: Optional[str] = None) -> str:
    if isinstance(dt, datetime.datetime):
        return dt.strftime(format)
    
    dt_datetime = new DatetimeParser().parse(...) if input_format else _str_to_datetime(dt)
    if format == "%s":
        return str(int(dt_datetime.timestamp()))
    return dt_datetime.strftime(format)
```

If I'm not mistaken with that, you will not need to chain anything